In [3]:
import math
from random import seed, randint
from copy import deepcopy
import time

def merge(a, b):
    if len(a) == 0:
        return b
    if len(b) == 0:
        return a
    result = []
    index_a = index_b = 0
    while len(result) < len(a) + len(b):
        if a[index_a] <= b[index_b]:
            result.append(a[index_a])
            index_a += 1
        else:
            result.append(b[index_b])
            index_b += 1
        if index_b == len(b):
            result += a[index_a:]
            break
        if index_a == len(a):
            result += b[index_b:]
            break
    return result

def merge_sort(a):
    if len(a)<=1:
        return a
    mid = len(a)//2
    left = merge_sort(a[:mid])
    right = merge_sort(a[mid:])
    return merge(left, right)      

def find_percentile(a, b, p):
    array = merge (a, b)
    c = merge_sort(array)
    if c == []:
        return None
    K = len(c)
    rank = math.ceil(p/100*K)-1
    if rank<0:
        return c[0]  
    return c[rank]

In [4]:
def test_find_percentile(a, b, p, correct_answer):
    percent_test = find_percentile(a, b, p)
    error_str = 'Test failled!'
    assert percent_test == correct_answer, error_str.format (a, b, p, percent_test, correct_answer)    

def run_unit_tests():
    test_find_percentile ([],[],10, None)
    test_find_percentile ([],[1],30,1)
    test_find_percentile ([1],[1],50,1)
    test_find_percentile ([1],[1,2,3],30,1)
    test_find_percentile ([3,4,5],[1,2,3],30,2)
    test_find_percentile ([1,2,5],[1000,1000,1000],50,5)
    test_find_percentile ([1,2,5],[1000,1000,1000,1000],50,1000)
    test_find_percentile ([0,34,57,18,21,44],[57,16,21,1000, 328],40,21)
    print('Unit test passed!')
    
def get_random_test (test_size, right_border):
    test = []
    for i in range(test_size):
        test.append(randint(0, right_border))  
    return test  

def reference_solution(input_a, input_b, input_p):
    def merge_test(a, b):
        c = []
        i, j = 0, 0
        while i < len(a) and j < len(b):
            if a[i] <= b[j]:
                c.append(a[i])
                i += 1
            else:
                c.append(b[j])
                j += 1
        while i < len(a):
            c.append(a[i])
            i += 1
        while j < len(b):
            c.append(b[j])
            j += 1
        return c

    a = deepcopy (input_a)
    b = deepcopy (input_b)
    p = deepcopy (input_p)
    result_array = merge_sort(merge_test(a,b))
    
    for i in range (len(a)):
        min_index = i+result_array[i:].index(min(result_array[i:]))
        result_array[i], result_array [min_index] = result_array [min_index], result_array [i]
        
    rank = math.ceil(p/100*len(result_array))-1
    if rank<0:
        return result_array[0]    
    return result_array[rank]

def run_stress_test(max_test_size=10, max_attempts=1000, max_right_border=10):
    seed (100)
    for test_size in range (1, max_test_size):
        for right_border in range (0, max_right_border):
            for attempt in range (max_attempts):
                test_1 = get_random_test (test_size, right_border)
                test_2 = get_random_test (test_size, right_border)
                percentil = randint (0, 100)
                test_find_percentile(test_1,test_2,percentil,reference_solution(test_1,test_2,percentil))
    print ('Stress test passed!')   
    
def run_max_test():
    seed(100)
    test_1 = get_random_test (15000, 1000000)
    test_2 = get_random_test (15000, 1000000)
    percentil = randint (0, 100)
    test_find_percentile(test_1,test_2,percentil,reference_solution(test_1,test_2,percentil))
    print ('Max test passed!')
    
    start = time.time()
    find_percentile(test_1,test_2,percentil)
    end = time.time()
    print ('Working time for find_percentile function:', end-start)

def main():
    run_unit_tests()
    run_stress_test()
    run_max_test()
    
if __name__=="__main__":
    main()

Unit test passed!
Stress test passed!
Max test passed!
Working time for find_percentile function: 0.2529621124267578


Working time for find_percentile function (max test): 0.2529621124267578. 

Merge() function has a O(n) time complexity, where n is a combined length of 2 input arrays. 
This function takes two arrays, the total length of which does not exceed n, and combines the two arrays, checking each element of which is at most once.

Merge_sort() function separates the input array recursively and calls the merge(). The number of steps is log 2 (n). As a result, the total time compexity is O (n* log 2 (n)), because the merge() is called for each half. 

Find_percentile() function takes an array as an input and returns element of rank, calculated according to the length of input array. Time complexity of this function is O(1).

The whole program requires an additional array sorted to store the resultant sorted array.

The time complexity = O (n* log 2 (n). 
The space complexity = O(n). 